wandb/sklearn/utils.py

```diff
@@ -0,0 +1,602 @@
+                        ):
+                            self.decision_boundary_points.append(db_point)
+                            self.decision_boundary_points_2d.append(db_point2d)
+                            break
+                        else:
+                            if self.verbose and not suppress_output:
+                                msg = "{} {}/{}: Rejected decision boundary keypoint (outside of plot area)"
+
+    def _find_decision_boundary_along_line(
+        self,
+        from_point,
+        to_point,
+        penalize_extremes=False,
+        penalize_tangent_distance=False,
+    ):
+        def objective(l, grad=0):
+            # interpolate between source and destionation; calculate distance from
+            # decision boundary
+            X = from_point + l[0] * (to_point - from_point)
+            error = self.decision_boundary_distance(X)
+
+            if penalize_tangent_distance:
+                # distance from tangent between class1 and class0 point in 2d space
+                x0, y0 = dimensionality_reduction.transform([X])[0]
+                x1, y1 = dimensionality_reduction.transform([from_point])[0]
+                x2, y2 = dimensionality_reduction.transform([to_point])[0]
+                error += (
+                    1e-12
+                    * np.abs((y2 - y1) * x0 - (x2 - x1) * y0 + x2 * y1 - y2 * x1)
+                    / np.sqrt((y2 - y1) ** 2 + (x2 - x1) ** 2)
+                )
+
+            if penalize_extremes:
+                error += 1e-8 * np.abs(0.5 - l[0])
+
+            return error
+
+        optimizer = self._get_optimizer()
+        optimizer.set_min_objective(objective)
+        cl = optimizer.optimize([random.random()])
+        db_point = from_point + cl[0] * (to_point - from_point)
+        return db_point
+
+    def _find_decision_boundary_on_hypersphere(self, centroid, R, penalize_known=False):
+        def objective(phi, grad=0):
+            # search on hypersphere surface in polar coordinates - map back to cartesian
+            cx = centroid + polar_to_cartesian(phi, R)
+            try:
+                cx2d = dimensionality_reduction.transform([cx])[0]
+                error = self.decision_boundary_distance(cx)
+                if penalize_known:
+                    # slight penalty for being too close to already known decision boundary
+                    # keypoints
+                    db_distances = [
+                        euclidean(cx2d, self.decision_boundary_points_2d[k])
+                        for k in range(len(self.decision_boundary_points_2d))
+                    ]
+                    error += (
+                        1e-8
+                        * (
+                            (self.mean_2d_dist - np.min(db_distances))
+                            / self.mean_2d_dist
+                        )
+                        ** 2
+                    )
+                return error
+            except (Exception, ex):
```

Comment:
This syntax is incorrect, it should be:
`except Exception as ex:`

As written:
it will catch either the exception "Exception" or ex (which isnt defined so it will throw an exception looking for an already defined symbol "ex" if an exception is ever raised)

Reply:
These errors are in the unittest failures in circleci:
```
============= 361 passed, 13 skipped, 5 warnings in 77.43 seconds ==============
./wandb/sklearn/utils.py:535:32: F821 undefined name 'ex'
./wandb/sklearn/utils.py:536:65: F821 undefined name 'ex'
./wandb/sklearn/utils.py:557:9: F821 undefined name 'opt'
./wandb/sklearn/utils.py:558:9: F821 undefined name 'opt'
./wandb/sklearn/utils.py:559:9: F821 undefined name 'opt'
./wandb/sklearn/utils.py:560:9: F821 undefined name 'opt'
./wandb/sklearn/utils.py:562:16: F821 undefined name 'opt'
```

Reply:
thank you, fixed [here](https://github.com/wandb/client/pull/719/commits/fdd28b9e5049bf95f6a4aab00ef316066bf9df84).